File: devices/sdk/kotlin/src/main/kotlin/ai/nexting/devices/DeviceInfo.kt
package ai.nexting.devices

import java.nio.ByteBuffer
import java.nio.charset.CodingErrorAction
import java.nio.charset.StandardCharsets
import java.util.UUID
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.longOrNull

data class DeviceIdentity(
    val deviceId: UUID?,
    val manufacturer: String?,
    val displayName: String?,
    val serialNumber: String?,
)

data class DeviceDisplay(
    val type: String,
    val width: Int,
    val height: Int,
)

data class DeviceCapabilities(
    val buttonCount: Int?,
    val approvalButtonCount: Int?,
    val customButtonCount: Int?,
    val rotaryCount: Int?,
    val rotaryPressCount: Int?,
    val statusSlots: Int,
    val batteryService: Boolean,
    val display: DeviceDisplay?,
    val haptics: List<String>,
)

data class DeviceVendorFact(
    val key: String,
    val label: String,
    val value: String,
)

data class DeviceVendorInfo(
    val namespace: String,
    val facts: List<DeviceVendorFact>,
)

data class DeviceInfo(
    val protocolName: String,
    val spec: String,
    val wireVersions: List<Int>,
    val profiles: List<String>,
    val model: String,
    val firmwareVersion: String,
    val maxMessageBytes: Long,
    val maxSummaryBytes: Int,
    val identity: DeviceIdentity,
    val capabilities: DeviceCapabilities,
    val vendor: DeviceVendorInfo?,
) {
    val supportsApprovalV1: Boolean
        get() = protocolName == "nexting-device" &&
            wireVersions.contains(1) &&
            profiles.contains("approval/1")

    val supportsStatusV1: Boolean
        get() = supportsApprovalV1 &&
            profiles.contains("status/1") &&
            capabilities.statusSlots > 0

    fun supportsProfile(profile: String): Boolean =
        supportsApprovalV1 && profiles.contains(profile)

    val supportsNavigationV1: Boolean
        get() = supportsProfile(DeviceMessageCodec.NAVIGATION_PROFILE)
    val supportsKeysV1: Boolean
        get() = supportsProfile(DeviceMessageCodec.KEYS_PROFILE)
    val supportsRotaryV1: Boolean
        get() = supportsProfile(DeviceMessageCodec.ROTARY_PROFILE)
    val supportsVoiceV1: Boolean
        get() = supportsProfile(DeviceMessageCodec.VOICE_PROFILE)
    val supportsTextV1: Boolean
        get() = supportsProfile(DeviceMessageCodec.TEXT_PROFILE)
    val supportsUsageV1: Boolean
        get() = supportsProfile(DeviceMessageCodec.USAGE_PROFILE)
    val supportsConfigV1: Boolean
        get() = supportsProfile(DeviceMessageCodec.CONFIG_PROFILE)
}

object DeviceBattery {
    const val SERVICE_UUID = "0000180f-0000-1000-8000-00805f9b34fb"
    const val LEVEL_CHARACTERISTIC_UUID = "00002a19-0000-1000-8000-00805f9b34fb"

    fun decodeLevel(bytes: ByteArray): Int? {
        if (bytes.size != 1) return null
        return (bytes[0].toInt() and 0xff).coerceAtMost(100)
    }
}

object DeviceInfoCodec {
    const val MAX_ENCODED_BYTES = 4_096
    private const val MAX_VENDOR_BYTES = 1_024
    private val json = Json {
        ignoreUnknownKeys = true
        isLenient = false
    }
    private val canonicalUnsigned = Regex("0|[1-9][0-9]*")
    private val uuid = Regex(
        "[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[1-5][0-9a-fA-F]{3}-" +
            "[89abAB][0-9a-fA-F]{3}-[0-9a-fA-F]{12}",
    )
    private val namespace = Regex(
        "[a-z0-9](?:[a-z0-9-]{0,62}\\.)+[a-z0-9][a-z0-9-]{0,62}",
    )
    private val factKey = Regex("[A-Za-z0-9][A-Za-z0-9._-]{0,31}")
    private val inertMarkup = Regex(
        "(?:</?[a-z]|https?://|www\\.|[`*_#\\[\\]()])",
        RegexOption.IGNORE_CASE,
    )

    fun decode(bytes: ByteArray): DeviceInfo? {
        if (bytes.size > MAX_ENCODED_BYTES) return null
        val text = decodeUtf8(bytes) ?: return null
        val root = runCatching { json.parseToJsonElement(text) }.getOrNull()
            as? JsonObject ?: return null
        if (!boundedDepth(root, 0)) return null

        val protocolName = root.requiredString("protocol") ?: return null
        val spec = root.requiredString("spec") ?: return null
        val model = root.requiredString("model") ?: return null
        val firmware = root.requiredString("fw") ?: return null
        if (
            protocolName != "nexting-device" ||
            !validText(spec) ||
            !validText(model) ||
            !validText(firmware)
        ) return null

        val wire = root.requiredArray("wire")
            ?.mapNotNull { canonicalInt(it, 1, 65_535) }
            ?: return null
        if (wire.isEmpty() || wire.size > 4 || wire.distinct().size != wire.size || 1 !in wire) {
            return null
        }

        val profiles = root.requiredArray("profiles")
            ?.mapNotNull {
                (it as? JsonPrimitive)
                    ?.takeIf(JsonPrimitive::isString)
                    ?.content
            }
            ?: return null
        if (
            profiles.isEmpty() ||
            profiles.size > 16 ||
            profiles.distinct().size != profiles.size ||
            profiles.any { !validText(it, 32) } ||
            "approval/1" !in profiles
        ) return null

        val maxMessageBytes = canonicalLong(root["max_message_bytes"], 512, UInt.MAX_VALUE.toLong())
            ?: return null
        val maxSummaryBytes = canonicalInt(root["max_summary_bytes"], 1, 240)
            ?: return null

        val statusSlots = optionalInt(root, "statusSlots", 0, 8) ?: return null
        val buttonCount = optionalInt(root, "button_count", 0, 1_024) ?: return null
        val approvalButtonCount =
            optionalInt(root, "approval_button_count", 0, 1_024) ?: return null
        val customButtonCount =
            optionalInt(root, "custom_button_count", 0, 1_024) ?: return null
        val rotaryCount = optionalInt(root, "rotary_count", 0, 64) ?: return null
        val rotaryPressCount =
            optionalInt(root, "rotary_press_count", 0, 64) ?: return null
        if (
            (statusSlots.value ?: 0) > 0 && "status/1" !in profiles ||
            !fits(approvalButtonCount.value, buttonCount.value) ||
            !fits(customButtonCount.value, buttonCount.value) ||
            !fits(rotaryPressCount.value, rotaryCount.value)
        ) return null

        val batteryService = when (val raw = root["battery_service"]) {
            null -> false
            is JsonPrimitive -> raw.booleanOrNull ?: return null
            else -> return null
        }
        val display = decodeDisplay(root["display"]) ?: if ("display" in root) return null else null
        val haptics = decodeHaptics(root["haptics"]) ?: if ("haptics" in root) return null else emptyList()

        val rawDeviceId = root.optionalString("device_id") ?: if ("device_id" in root) return null else null
        val deviceId = if (rawDeviceId != null) {
            if (!uuid.matches(rawDeviceId)) return null
            runCatching { UUID.fromString(rawDeviceId) }.getOrNull() ?: return null
        } else {
            null
        }
        val manufacturer = validatedOptionalText(root, "manufacturer") ?: if ("manufacturer" in root) return null else null
        val displayName = validatedOptionalText(root, "display_name") ?: if ("display_name" in root) return null else null
        val serialNumber = validatedOptionalText(root, "serial_number") ?: if ("serial_number" in root) return null else null

        return DeviceInfo(
            protocolName = protocolName,
            spec = spec,
            wireVersions = wire,
            profiles = profiles,
            model = model,
            firmwareVersion = firmware,
            maxMessageBytes = maxMessageBytes,
            maxSummaryBytes = maxSummaryBytes,
            identity = DeviceIdentity(deviceId, manufacturer, displayName, serialNumber),
            capabilities = DeviceCapabilities(
                buttonCount = buttonCount.value,
                approvalButtonCount = approvalButtonCount.value,
                customButtonCount = customButtonCount.value,
                rotaryCount = rotaryCount.value,
                rotaryPressCount = rotaryPressCount.value,
                statusSlots = statusSlots.value ?: 0,
                batteryService = batteryService,
                display = display,
                haptics = haptics,
            ),
            vendor = decodeVendor(root["vendor"]),
        )
    }

    private fun decodeUtf8(bytes: ByteArray): String? = runCatching {
        StandardCharsets.UTF_8.newDecoder()
            .onMalformedInput(CodingErrorAction.REPORT)
            .onUnmappableCharacter(CodingErrorAction.REPORT)
            .decode(ByteBuffer.wrap(bytes))
            .toString()
    }.getOrNull()

    private fun boundedDepth(value: JsonElement, depth: Int): Boolean {
        if (depth > 8) return false
        return when (value) {
            is JsonObject -> value.values.all { boundedDepth(it, depth + 1) }
            is JsonArray -> value.all { boundedDepth(it, depth + 1) }
            else -> true
        }
    }

    private fun validText(value: String, maximumBytes: Int = 64): Boolean =
        value.isNotEmpty() &&
            value.encodeToByteArray().size <= maximumBytes &&
            value.none { it.code < 0x20 || it.code == 0x7f }

    private fun validatedOptionalText(root: JsonObject, key: String): String? =
        root.optionalString(key)?.takeIf(::validText)

    private fun canonicalInt(value: JsonElement?, minimum: Int, maximum: Int): Int? {
        val primitive = value as? JsonPrimitive ?: return null
        if (primitive.isString || !canonicalUnsigned.matches(primitive.content)) return null
        return primitive.intOrNull?.takeIf { it in minimum..maximum }
    }

    private fun canonicalLong(value: JsonElement?, minimum: Long, maximum: Long): Long? {
        val primitive = value as? JsonPrimitive ?: return null
        if (primitive.isString || !canonicalUnsigned.matches(primitive.content)) return null
        return primitive.longOrNull?.takeIf { it in minimum..maximum }
    }

    private data class OptionalInt(val value: Int?)

    private fun optionalInt(
        root: JsonObject,
        key: String,
        minimum: Int,
        maximum: Int,
    ): OptionalInt? {
        val raw = root[key] ?: return OptionalInt(null)
        return canonicalInt(raw, minimum, maximum)?.let { OptionalInt(it) }
    }

    private fun fits(value: Int?, total: Int?): Boolean =
        value == null || total == null || value <= total

    private fun decodeDisplay(raw: JsonElement?): DeviceDisplay? {
        val display = raw as? JsonObject ?: return null
        val type = display.requiredString("type")?.takeIf { validText(it, 32) } ?: return null
        val width = canonicalInt(display["width"], 1, 4_096) ?: return null
        val height = canonicalInt(display["height"], 1, 4_096) ?: return null
        return DeviceDisplay(type, width, height)
    }

    private fun decodeHaptics(raw: JsonElement?): List<String>? {
        val values = raw as? JsonArray ?: return null
        val decoded = values.mapNotNull {
            (it as? JsonPrimitive)
                ?.takeIf(JsonPrimitive::isString)
                ?.content
        }
        if (
            decoded.isEmpty() ||
            decoded.size > 8 ||
            decoded.size != values.size ||
            decoded.distinct().size != decoded.size ||
            decoded.any { !validText(it, 32) }
        ) return null
        return decoded
    }

    private fun decodeVendor(raw: JsonElement?): DeviceVendorInfo? {
        val vendor = raw as? JsonObject ?: return null
        if (vendor.toString().encodeToByteArray().size > MAX_VENDOR_BYTES) return null
        val vendorNamespace = vendor.requiredString("namespace")
            ?.takeIf { validText(it, 128) && namespace.matches(it) }
            ?: return null
        val rawFacts = vendor["facts"] as? JsonArray ?: return null
        if (rawFacts.size !in 1..16) return null
        val seen = mutableSetOf<String>()
        val facts = rawFacts.map { rawFact ->
            val fact = rawFact as? JsonObject ?: return null
            val key = fact.requiredString("key")
                ?.takeIf { factKey.matches(it) && seen.add(it) }
                ?: return null
            val label = fact.requiredString("label")
                ?.takeIf { validInertText(it, 64) }
                ?: return null
            val value = vendorValue(fact["value"])
                ?.takeIf { validInertText(it, 128) }
                ?: return null
            DeviceVendorFact(key, label, value)
        }
        return DeviceVendorInfo(vendorNamespace, facts)
    }

    private fun vendorValue(raw: JsonElement?): String? {
        val primitive = raw as? JsonPrimitive ?: return null
        if (primitive == JsonNull || primitive.booleanOrNull != null) return null
        if (primitive.isString) return primitive.content
        if (!canonicalUnsigned.matches(primitive.content)) return null
        return primitive.longOrNull?.toString()
    }

    private fun validInertText(value: String, maximumBytes: Int): Boolean =
        validText(value, maximumBytes) && !inertMarkup.containsMatchIn(value)

    private fun JsonObject.requiredString(key: String): String? {
        val primitive = this[key] as? JsonPrimitive ?: return null
        return primitive.contentOrNull.takeIf { primitive.isString }
    }

    private fun JsonObject.optionalString(key: String): String? =
        if (key !in this) null else requiredString(key)

    private fun JsonObject.requiredArray(key: String): JsonArray? = this[key] as? JsonArray

    private val JsonPrimitive.contentOrNull: String?
        get() = if (this == JsonNull) null else content
}

File: devices/sdk/swift/Sources/NextingDeviceKit/DeviceInfo.swift
import Foundation

public struct NextingDeviceIdentity: Equatable, Sendable {
    public let deviceId: UUID?
    public let manufacturer: String?
    public let displayName: String?
    public let serialNumber: String?
}

public struct NextingDeviceDisplay: Equatable, Sendable {
    public let type: String
    public let width: Int
    public let height: Int
}

public struct NextingDeviceCapabilities: Equatable, Sendable {
    public let buttonCount: Int?
    public let approvalButtonCount: Int?
    public let customButtonCount: Int?
    public let rotaryCount: Int?
    public let rotaryPressCount: Int?
    public let statusSlots: Int
    public let batteryService: Bool
    public let display: NextingDeviceDisplay?
    public let haptics: [String]
}

public struct NextingDeviceVendorFact: Equatable, Sendable {
    public let key: String
    public let label: String
    public let value: String
}

public struct NextingDeviceVendorInfo: Equatable, Sendable {
    public let namespace: String
    public let facts: [NextingDeviceVendorFact]
}

public struct NextingDeviceInfo: Equatable, Sendable {
    public static let maxEncodedBytes = 4_096

    public let protocolName: String
    public let spec: String
    public let wireVersions: [Int]
    public let profiles: [String]
    public let model: String
    public let firmwareVersion: String
    public let maxMessageBytes: Int
    public let maxSummaryBytes: Int
    public let identity: NextingDeviceIdentity
    public let capabilities: NextingDeviceCapabilities
    public let vendor: NextingDeviceVendorInfo?

    public var statusSlots: Int { capabilities.statusSlots }

    public var supportsApprovalV1: Bool {
        protocolName == "nexting-device"
            && wireVersions.contains(NextingDeviceCodec.wireVersion)
            && profiles.contains(NextingDeviceCodec.profile)
    }

    public var supportsStatusV1: Bool {
        supportsApprovalV1
            && profiles.contains(NextingDeviceCodec.statusProfile)
            && statusSlots >= 1
    }

    public func supportsProfile(_ profile: String) -> Bool {
        supportsApprovalV1 && profiles.contains(profile)
    }

    public var supportsNavigationV1: Bool {
        supportsProfile(NextingDeviceCodec.navigationProfile)
    }

    public var supportsKeysV1: Bool {
        supportsProfile(NextingDeviceCodec.keysProfile)
    }

    public var supportsRotaryV1: Bool {
        supportsProfile(NextingDeviceCodec.rotaryProfile)
    }

    public var supportsVoiceV1: Bool {
        supportsProfile(NextingDeviceCodec.voiceProfile)
    }

    public var supportsTextV1: Bool {
        supportsProfile(NextingDeviceCodec.textProfile)
    }

    public var supportsUsageV1: Bool {
        supportsProfile(NextingDeviceCodec.usageProfile)
    }

    public var supportsConfigV1: Bool {
        supportsProfile(NextingDeviceCodec.configProfile)
    }

    public static func decode(_ data: Data) -> NextingDeviceInfo? {
        guard data.count <= maxEncodedBytes,
              String(data: data, encoding: .utf8) != nil,
              let payload = try? JSONDecoder().decode(Payload.self, from: data),
              payload.protocolName == "nexting-device",
              validDescriptor(payload.spec),
              validUniqueDescriptors(payload.profiles, maximumCount: 16, maximumBytes: 32),
              payload.profiles.contains(NextingDeviceCodec.profile),
              !payload.wireVersions.isEmpty,
              payload.wireVersions.count <= 4,
              Set(payload.wireVersions).count == payload.wireVersions.count,
              payload.wireVersions.allSatisfy({ (1 ... 65_535).contains($0) }),
              payload.wireVersions.contains(NextingDeviceCodec.wireVersion),
              validDescriptor(payload.model),
              validDescriptor(payload.firmwareVersion),
              (512 ... Int(UInt32.max)).contains(payload.maxMessageBytes),
              (1 ... NextingDeviceCodec.maxSummaryBytes).contains(payload.maxSummaryBytes)
        else { return nil }

        guard validOptionalDescriptor(payload.manufacturer),
              validOptionalDescriptor(payload.displayName),
              validOptionalDescriptor(payload.serialNumber)
        else { return nil }

        let statusSlots = payload.statusSlots ?? 0
        guard (0 ... NextingDeviceCodec.maxStatusAgents).contains(statusSlots),
              statusSlots == 0 || payload.profiles.contains(NextingDeviceCodec.statusProfile),
              validCount(payload.buttonCount, maximum: 1_024),
              validCount(payload.approvalButtonCount, maximum: 1_024),
              validCount(payload.customButtonCount, maximum: 1_024),
              validCount(payload.rotaryCount, maximum: 64),
              validCount(payload.rotaryPressCount, maximum: 64),
              specializedCount(payload.approvalButtonCount, fitsWithin: payload.buttonCount),
              specializedCount(payload.customButtonCount, fitsWithin: payload.buttonCount),
              specializedCount(payload.rotaryPressCount, fitsWithin: payload.rotaryCount),
              validDisplay(payload.display),
              validHaptics(payload.haptics)
        else { return nil }

        let deviceId: UUID?
        if let rawDeviceId = payload.deviceId {
            guard rawDeviceId.utf8.count == 36,
                  let parsed = UUID(uuidString: rawDeviceId)
            else { return nil }
            deviceId = parsed
        } else {
            deviceId = nil
        }

        let info = NextingDeviceInfo(
            protocolName: payload.protocolName,
            spec: payload.spec,
            wireVersions: payload.wireVersions,
            profiles: payload.profiles,
            model: payload.model,
            firmwareVersion: payload.firmwareVersion,
            maxMessageBytes: payload.maxMessageBytes,
            maxSummaryBytes: payload.maxSummaryBytes,
            identity: NextingDeviceIdentity(
                deviceId: deviceId,
                manufacturer: payload.manufacturer,
                displayName: payload.displayName,
                serialNumber: payload.serialNumber
            ),
            capabilities: NextingDeviceCapabilities(
                buttonCount: payload.buttonCount,
                approvalButtonCount: payload.approvalButtonCount,
                customButtonCount: payload.customButtonCount,
                rotaryCount: payload.rotaryCount,
                rotaryPressCount: payload.rotaryPressCount,
                statusSlots: statusSlots,
                batteryService: payload.batteryService ?? false,
                display: payload.display.map {
                    NextingDeviceDisplay(type: $0.type, width: $0.width, height: $0.height)
                },
                haptics: payload.haptics ?? []
            ),
            vendor: decodeVendor(from: data)
        )
        return info.supportsApprovalV1 ? info : nil
    }

    private static func validDescriptor(
        _ value: String,
        maximumBytes: Int = 64
    ) -> Bool {
        !value.isEmpty
            && value.utf8.count <= maximumBytes
            && value.unicodeScalars.allSatisfy {
                !CharacterSet.controlCharacters.contains($0)
            }
    }

    private static func validOptionalDescriptor(_ value: String?) -> Bool {
        value.map { validDescriptor($0) } ?? true
    }

    private static func validUniqueDescriptors(
        _ values: [String],
        maximumCount: Int,
        maximumBytes: Int
    ) -> Bool {
        !values.isEmpty
            && values.count <= maximumCount
            && Set(values).count == values.count
            && values.allSatisfy { validDescriptor($0, maximumBytes: maximumBytes) }
    }

    private static func validCount(_ value: Int?, maximum: Int) -> Bool {
        value.map { (0 ... maximum).contains($0) } ?? true
    }

    private static func specializedCount(_ value: Int?, fitsWithin total: Int?) -> Bool {
        guard let value, let total else { return true }
        return value <= total
    }

    private static func validDisplay(_ display: Payload.Display?) -> Bool {
        guard let display else { return true }
        return validDescriptor(display.type, maximumBytes: 32)
            && (1 ... 4_096).contains(display.width)
            && (1 ... 4_096).contains(display.height)
    }

    private static func validHaptics(_ values: [String]?) -> Bool {
        guard let values else { return true }
        return validUniqueDescriptors(values, maximumCount: 8, maximumBytes: 32)
    }

    private static func decodeVendor(from data: Data) -> NextingDeviceVendorInfo? {
        guard let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let rawVendor = root["vendor"]
        else { return nil }
        guard let vendor = rawVendor as? [String: Any],
              let namespace = vendor["namespace"] as? String,
              validDescriptor(namespace, maximumBytes: 128),
              namespace.range(
                of: #"^[a-z0-9](?:[a-z0-9-]{0,62}\.)+[a-z0-9][a-z0-9-]{0,62}$"#,
                options: .regularExpression
              ) != nil,
              let rawFacts = vendor["facts"] as? [[String: Any]],
              (1 ... 16).contains(rawFacts.count),
              let encodedVendor = try? JSONSerialization.data(
                withJSONObject: vendor,
                options: [.sortedKeys]
              ),
              encodedVendor.count <= 1_024
        else { return nil }

        var keys = Set<String>()
        var facts: [NextingDeviceVendorFact] = []
        for rawFact in rawFacts {
            guard let key = rawFact["key"] as? String,
                  key.range(
                    of: #"^[A-Za-z0-9][A-Za-z0-9._-]{0,31}$"#,
                    options: .regularExpression
                  ) != nil,
                  keys.insert(key).inserted,
                  let label = rawFact["label"] as? String,
                  validInertText(label, maximumBytes: 64),
                  let value = vendorValue(rawFact["value"]),
                  validInertText(value, maximumBytes: 128)
            else { return nil }
            facts.append(NextingDeviceVendorFact(key: key, label: label, value: value))
        }
        return NextingDeviceVendorInfo(namespace: namespace, facts: facts)
    }

    private static func vendorValue(_ raw: Any?) -> String? {
        if let value = raw as? String { return value }
        if raw is Bool { return nil }
        if let value = raw as? NSNumber {
            let double = value.doubleValue
            guard double.isFinite,
                  double.rounded(.towardZero) == double
            else { return nil }
            return value.stringValue
        }
        return nil
    }

    private static func validInertText(_ value: String, maximumBytes: Int) -> Bool {
        validDescriptor(value, maximumBytes: maximumBytes)
            && value.range(
                of: #"(?:</?[a-z]|https?://|www\.|[`*_#\[\]()])"#,
                options: [.regularExpression, .caseInsensitive]
            ) == nil
    }

    private struct Payload: Decodable {
        struct Display: Decodable {
            let type: String
            let width: Int
            let height: Int
        }

        let protocolName: String
        let spec: String
        let wireVersions: [Int]
        let profiles: [String]
        let model: String
        let firmwareVersion: String
        let maxMessageBytes: Int
        let maxSummaryBytes: Int
        let statusSlots: Int?
        let deviceId: String?
        let manufacturer: String?
        let displayName: String?
        let serialNumber: String?
        let buttonCount: Int?
        let approvalButtonCount: Int?
        let customButtonCount: Int?
        let rotaryCount: Int?
        let rotaryPressCount: Int?
        let batteryService: Bool?
        let display: Display?
        let haptics: [String]?

        enum CodingKeys: String, CodingKey {
            case protocolName = "protocol"
            case spec
            case wireVersions = "wire"
            case profiles
            case model
            case firmwareVersion = "fw"
            case maxMessageBytes = "max_message_bytes"
            case maxSummaryBytes = "max_summary_bytes"
            case statusSlots
            case deviceId = "device_id"
            case manufacturer
            case displayName = "display_name"
            case serialNumber = "serial_number"
            case buttonCount = "button_count"
            case approvalButtonCount = "approval_button_count"
            case customButtonCount = "custom_button_count"
            case rotaryCount = "rotary_count"
            case rotaryPressCount = "rotary_press_count"
            case batteryService = "battery_service"
            case display
            case haptics
        }
    }
}
